consul/client.go

```diff
@@ -369,77 +360,55 @@ func (c *Client) localEvent(event serf.UserEvent) {
 
 // RPC is used to forward an RPC call to a consul server, or fail if no servers
 func (c *Client) RPC(method string, args interface{}, reply interface{}) error {
-	// Check to make sure we haven't spent too much time querying a
-	// single server
-	now := time.Now()
-	if !c.connRebalanceTime.IsZero() && now.After(c.connRebalanceTime) {
-		c.logger.Printf("[DEBUG] consul: connection time to server %s exceeded, rotating server connection", c.lastServer.Addr)
-		c.lastServer = nil
-	}
-
-	// Allocate these vars on the stack before the goto
-	var numConsulServers int
-	var clusterWideRebalanceConnsPerSec float64
-	var connReuseLowWaterMark time.Duration
-	var numLANMembers int
-
-	// Check the last RPC time, continue to reuse cached connection for
-	// up to clientRPCMinReuseDuration unless exceeded
-	// clientRPCConnMaxIdle
-	lastRPCTime := now.Sub(c.lastRPCTime)
-	var server *serverParts
-	if c.lastServer != nil && lastRPCTime < clientRPCConnMaxIdle {
-		server = c.lastServer
-		goto TRY_RPC
+	serverCfgPtr := c.serverConfigValue.Load()
+	if serverCfgPtr == nil {
+		c.logger.Printf("[ERR] consul: Failed to load a server config")
+		return structs.ErrNoServers
 	}
+	serverCfg := serverCfgPtr.(serverConfig)
 
-	// Bail if we can't find any servers
-	c.consulLock.RLock()
-	numConsulServers = len(c.consuls)
-	if numConsulServers == 0 {
-		c.consulLock.RUnlock()
+	numServers := len(serverCfg.servers)
+	if numServers == 0 {
+		c.logger.Printf("[ERR] consul: No servers found in the server config")
 		return structs.ErrNoServers
 	}
 
-	// Select a random addr
-	server = c.consuls[rand.Int31n(int32(numConsulServers))]
-	c.consulLock.RUnlock()
-
-	// Limit this connection's life based on the size (and health) of the
-	// cluster.  Never rebalance a connection more frequently than
-	// connReuseLowWaterMark, and make sure we never exceed
-	// clusterWideRebalanceConnsPerSec operations/s across numLANMembers.
-	clusterWideRebalanceConnsPerSec = float64(numConsulServers * newRebalanceConnsPerSecPerServer)
-	connReuseLowWaterMark = clientRPCMinReuseDuration + lib.RandomStagger(clientRPCMinReuseDuration/clientRPCJitterFraction)
-	numLANMembers = len(c.LANMembers())
-	c.connRebalanceTime = now.Add(lib.RateScaledInterval(clusterWideRebalanceConnsPerSec, connReuseLowWaterMark, numLANMembers))
-	c.logger.Printf("[DEBUG] consul: connection to server %s will expire at %v", server.Addr, c.connRebalanceTime)
+	// Find the first non-failing server in the server list.  If this is
+	// not the first server a prior RPC call marked the first server as
+	// failed and we're waiting for the server management task to reorder
+	// a working server to the front of the list.
+	var server *serverParts
+	for i := range serverCfg.servers {
+		failCount := atomic.LoadUint64(&(serverCfg.servers[i].Disabled))
```

Comment:
I don't think the copy buys you much in this case, though, since we don't hold on to it for very long. In that example it looks like the benefit is that the copy is detached and will age out, but we don't really need that behavior here.